### PR TITLE
Make EnsureService to do ports check if server spec and also to allow…

### PR DESCRIPTION
**Pull request by ycliuhw**
## Description of change

Make EnsureService:

- do ports check if server spec;
- allow headless service has storage config

## QA steps

- deploy charm with omitServiceFrontend: false && no ports specified in containers;

```yaml
omitServiceFrontend: false
containers:
  - name: %(name)s
    imageDetails:
      imagePath: %(docker_image_path)s
      username: %(docker_image_username)s
      password: %(docker_image_password)s
    # ports:
    # - containerPort: %(port)s
    #   protocol: TCP
...
```

```
# got error: 
ports are required for kubernetes service
```

- deploy charm with omitServiceFrontend: true && no ports specified in containers - no errors;

```yaml
omitServiceFrontend: true
containers:
  - name: %(name)s
    imageDetails:
      imagePath: %(docker_image_path)s
      username: %(docker_image_username)s
      password: %(docker_image_password)s
    # ports:
    # - containerPort: %(port)s
    #   protocol: TCP
```

```bash
$ mkubectl get all,pv,pvc,cm,po,secrets,ing,endpoints -o wide -n t1

NAME                         READY   STATUS    RESTARTS   AGE     IP          NODE                   NOMINATED NODE   READINESS GATES
pod/mariadb-k8s-0            1/1     Running   0          2m29s   10.1.1.64   kelvinliu-virtualbox   <none>           <none>
pod/mariadb-k8s-operator-0   1/1     Running   0          2m37s   10.1.1.63   kelvinliu-virtualbox   <none>           <none>

NAME                            TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)   AGE     SELECTOR
service/mariadb-k8s-endpoints   ClusterIP   None         <none>        <none>    2m29s   juju-app=mariadb-k8s
# no service created for mariadb.

NAME                                    READY   AGE     CONTAINERS      IMAGES
statefulset.apps/mariadb-k8s            1/1     2m29s   mariadb-k8s     mariadb
statefulset.apps/mariadb-k8s-operator   1/1     2m37s   juju-operator   ycliuhw/jujud-operator:2.7-beta1

NAME                                                        CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM                                STORAGECLASS        REASON   AGE     VOLUMEMODE
persistentvolume/pvc-e8217fdb-6ba9-45aa-b46a-d755427375f2   1Gi        RWO            Delete           Bound    t1/database-1cf00409-mariadb-k8s-0   microk8s-hostpath            2m29s   Filesystem
persistentvolume/pvc-fe742f7b-4630-4dbe-ac44-d263460c3779   1Gi        RWO            Delete           Bound    t1/charm-mariadb-k8s-operator-0      microk8s-hostpath            2m36s   Filesystem

NAME                                                    STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS        AGE     VOLUMEMODE
persistentvolumeclaim/charm-mariadb-k8s-operator-0      Bound    pvc-fe742f7b-4630-4dbe-ac44-d263460c3779   1Gi        RWO            microk8s-hostpath   2m37s   Filesystem
persistentvolumeclaim/database-1cf00409-mariadb-k8s-0   Bound    pvc-e8217fdb-6ba9-45aa-b46a-d755427375f2   1Gi        RWO            microk8s-hostpath   2m29s   Filesystem

NAME                                          DATA   AGE
configmap/mariadb-k8s-configurations-config   1      2m29s
configmap/mariadb-k8s-operator-config         1      2m37s

NAME                         TYPE                                  DATA   AGE
secret/default-token-tc2c2   kubernetes.io/service-account-token   3      9m32s


NAME                              ENDPOINTS   AGE
endpoints/mariadb-k8s-endpoints   10.1.1.64   2m29s

```

## Documentation changes

No.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1838664

### CODE REVIEW

File: caas/kubernetes/provider/k8s.go [hunk 1]
```diff
@@ -1118,9 +1118,6 @@ func (k *kubernetesClient) EnsureService(
 	if params == nil || params.PodSpec == nil {
 		return errors.Errorf("missing pod spec")
 	}
-	if params.PodSpec.OmitServiceFrontend && len(params.Filesystems) == 0 {
-		return errors.Errorf("kubernetes service is required when using storage")
```
Comment (ycliuhw) [inline caas/kubernetes/provider/k8s.go]: what's the initial purpose to do this check here?

**Comment by ycliuhw** [comment]
[https://bugs.launchpad.net/juju/+bug/1838664](https://trello.com/c/sOASuIsJ/872-https-bugslaunchpadnet-juju-bug-1838664)

File: caas/kubernetes/provider/k8s.go (hunk 1 above)
Comment (wallyworld) [reply caas/kubernetes/provider/k8s.go]: not sure anymore

**Comment by ycliuhw** [comment]
as discussed,  we still need `omitServiceFrontend, so keep it for now.

**Comment by ycliuhw** [comment]
$$merge$$

**Pull request merged**